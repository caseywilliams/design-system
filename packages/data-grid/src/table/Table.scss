@import '@puppet/sass-variables/_index';

.rc-table-container {
  overflow-x: auto;
}

.dg-empty-state-container {
  margin: auto;
  text-align: center;
  padding-top: 100px;
  padding-bottom: 100px;

  .dg-empty-state-message {
    margin-top: 8px;
  }
}

.dg-column-header-label-active {
  color: $puppet-b500;
  width: fit-content;
  margin: 0;
  padding: 0;
}

.dg-column-header-icon-container {
  float: right;
  display: flex;
  flex-direction: column;
  font-size: 7px;
  outline: none;
  cursor: pointer;
}

.dg-column-header-icon-up {
  outline: none;
}
.dg-column-header-icon-down {
  outline: none;
}

.asc {
  .dg-column-header-icon-up {
    color: $puppet-b500;
    outline: none;
  }
}
.desc {
  outline: none;
  .dg-column-header-icon-down {
    color: $puppet-b500;
    outline: none;
  }
}

.dg-table-horizontal-scroll {
  margin: auto;
  overflow: auto;
}

.dg-table-fixed-column table {
  width: 100%;
  border-collapse: separate;
  border-spacing: 0;

  th:first-child {
    position: -webkit-sticky;
    position: sticky;
    left: 0;
    z-index: 2;
  }

  td:first-child {
    position: -webkit-sticky;
    position: sticky;
    left: 0;
    z-index: 2;
    background: #ccc;
  }
}

.dg-table-header-checkbox {
  width: 20px;
}

.dg-table {
  background-color: $puppet-white;
  border-collapse: collapse;
  border-spacing: 0;
  table-layout: auto;
  width: 100%;
}

.dg-table-cell,
.dg-table-header-cell {
  position: relative;

  &:not(:first-child) {
    padding-left: 18px;
  }

  &:not(:first-child)::before {
    background-color: $puppet-white;
    bottom: 0;
    content: '';
    left: 0;
    position: absolute;
    top: 0;
    width: 2px;
  }
}

.dg-table-header-cell {
  @include puppet-type-h6(medium);
  background-color: $puppet-n200;
  padding: 11px 16px;
  text-align: left;
}

.dg-table-cell {
  @include puppet-type-small();
  padding: 15px 16px;
  line-height: 18px;
}

tr {
  border-bottom: $puppet-common-border;
}
