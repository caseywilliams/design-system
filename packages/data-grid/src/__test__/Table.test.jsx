import React from 'react';
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Table from '../table/Table';

configure({ adapter: new Adapter() });

const data = [
  { id: 1, eventType: 'Virus/Malware', affectedDevices: 20, detections: 634 },
  {
    id: 2,
    eventType: 'Spyware/Grayware',
    affectedDevices: 20,
    detections: 634,
  },
  { id: 3, eventType: 'URL Filtering', affectedDevices: 15, detections: 598 },
  { id: 4, eventType: 'Web Reputation', affectedDevices: 15, detections: 598 },
  { id: 5, eventType: 'Network Virus', affectedDevices: 15, detections: 497 },
  {
    id: 6,
    eventType: 'Application Control',
    affectedDevices: 0,
    detections: 0,
  },
  {
    id: 7,
    eventType: 'Application Control',
    affectedDevices: 0,
    detections: 0,
  },
];

const columns = [
  { label: 'Event Type', dataKey: 'eventType' },
  { label: 'Affected Devices', dataKey: 'affectedDevices' },
  { label: 'Detections', dataKey: 'detections' },
  { label: 'Linked field', dataKey: 'Link' },
];

const wrapper = mount(<Table columns={columns} data={data} />);
const wrapper2 = <Table columns={columns} />;
const wrapper3 = mount(<Table columns={columns} data={[]} />);
const wrapper4 = mount(<Table columns={columns} data={data} sortable />);
const wrapper6 = mount(
  <Table
    columns={columns}
    data={data}
    rowClassName="testRowClassName"
    columnClassName="testColumnClassName"
  />,
);
const wrapper10 = mount(<Table columns={columns} data={data} selectable />);

describe('Snapshot test', () => {
  test('Check component matches previous HTML snapshot', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper3).toMatchSnapshot();
  });
});

describe('Data Prop', () => {
  test('Check data prop default value', () => {
    const props = mount(wrapper2).props();
    expect(props.data).toEqual([]);
  });

  test('Empty array should render empty state', () => {
    expect(
      wrapper3
        .find('div.dg-empty-state-container')
        .hasClass('dg-empty-state-container'),
    ).toEqual(true);
  });

  test('Table should render correct number of rows', () => {
    expect(wrapper.find('tr.dg-table-row')).toHaveLength(7);
  });

  test('Data values should render in correct column', () => {
    expect(
      wrapper
        .find('tr.dg-table-row-0')
        .contains(<td className="rc-table-cell">Virus/Malware</td>),
    );
    expect(
      wrapper
        .find('tr.dg-table-row-0')
        .contains(<td className="rc-table-cell">20</td>),
    );
    expect(
      wrapper
        .find('tr.dg-table-row-0')
        .contains(<td className="rc-table-cell">634</td>),
    );
  });
});

describe('Column Prop', () => {
  test('All columns render', () => {
    expect(wrapper.find('tr.rc-table-header').children()).toHaveLength(4);
  });

  test('Correct labels displayed / Columns render in correct order', () => {
    expect(
      wrapper
        .find('tr.rc-table-header')
        .childAt(0)
        .text(),
    ).toBe(columns[0].label);

    expect(
      wrapper
        .find('tr.rc-table-header')
        .childAt(1)
        .text(),
    ).toBe(columns[1].label);

    expect(
      wrapper
        .find('tr.rc-table-header')
        .childAt(2)
        .text(),
    ).toBe(columns[2].label);

    expect(
      wrapper
        .find('tr.rc-table-header')
        .childAt(3)
        .text(),
    ).toBe(columns[3].label);
  });

  const data2 = [
    {
      eventType: 'Virus/Malware',
      eventResults: { affectedDevices: 20, detections: 600, unique: 40 },
    },
  ];

  const columns2 = [
    {
      label: 'Event Type1',
      dataKey: 'eventType',
    },
    {
      label: 'Affected Devices',
      dataKey: 'eventResults.affectedDevices',
    },
    { label: 'Detections', dataKey: 'eventResults.detections' },
    { label: 'Linked field', dataKey: 'eventResults.unique' },
  ];

  const wrapper8 = mount(<Table data={data2} columns={columns2} />);

  test('Customs can collect data with dataKey paths', () => {
    expect(
      wrapper8
        .find('td.rc-table-cell')
        .at(2)
        .text(),
    ).toEqual('600');
  });
});

describe('Sortable Props', () => {
  const mockfunc = jest.fn();
  const sortedState = { direction: 'asc', sortDataKey: 'eventType' };
  const wrapper5 = mount(
    <Table
      columns={columns}
      data={data}
      sortable
      onSort={mockfunc}
      sortedColumn={sortedState}
    />,
  );

  test('When sortable prop is true header icons render ', () => {
    expect(wrapper4.find('.rc-icon-increment')).toHaveLength(columns.length);
  });

  test('SortedColumn renders correctly', () => {
    expect(
      wrapper5
        .find('.rc-table-header')
        .children()
        .exists('.dg-column-header-label-active'),
    ).toBe(true);
  });
});

describe('Custom classes', () => {
  const rowMockfunc = jest.fn();
  const columnMockfunc = jest.fn();
  // eslint-disable-next-line
  const wrapper7 = mount(
    <Table
      columns={columns}
      data={data}
      rowClassName={rowMockfunc}
      columnClassName={columnMockfunc}
    />,
  );

  test('Custom row classname of string is rendered', () => {
    wrapper6.find('tr.dg-table-row').forEach(node => {
      expect(node.hasClass('testRowClassName')).toEqual(true);
    });
  });

  test('Custom row classname of function is called', () => {
    expect(rowMockfunc).toHaveBeenCalledTimes(7);
  });

  test('Custom column classname of string is rendered', () => {
    wrapper6.find('td.rc-table-cell').forEach(node => {
      expect(node.hasClass('testColumnClassName')).toEqual(true);
    });
  });

  test('Custom column classname of function is called', () => {
    expect(columnMockfunc).toHaveBeenCalledTimes(28);
  });
});

describe('Selection Props', () => {
  const headerCheckMockfunc = jest.fn();
  const rowCheckMockfunc = jest.fn();
  const wrapper9 = mount(
    <Table
      columns={columns}
      data={data}
      selectable
      onHeaderChecked={headerCheckMockfunc}
      headerCheckState
      onRowChecked={rowCheckMockfunc}
    />,
  );

  test('When selectable prop passed table renders correctly', () => {
    expect(
      wrapper9
        .find('.dg-table-header-checkbox')
        .first()
        .parent()
        .is('th.dg-table-header-checkbox-container'),
    ).toEqual(true);
  });
  expect(wrapper9.findWhere(n => n.name() === 'Checkbox')).toHaveLength(8);

  test('When header checkbox is clicked fire onHeaderChecked', () => {
    wrapper9
      .find('.rc-checkbox')
      .first()
      .simulate('change');
    expect(headerCheckMockfunc).toHaveBeenCalled();
  });

  test('When row checkbox is clicked fire onRowChecked', () => {
    wrapper9
      .find('.rc-checkbox')
      .last()
      .simulate('change');
    expect(rowCheckMockfunc).toHaveBeenCalled();
  });

  test('headerCheckState passes checkbox state correctly', () => {
    expect(
      wrapper9
        .find('.dg-table-header-checkbox')
        .first()
        .prop('checked'),
    ).toBeTruthy();

    expect(
      wrapper10
        .find('.dg-table-header-checkbox')
        .first()
        .prop('checked'),
    ).toBeFalsy();
  });

  test('When property `selected` is true in data prop object table row cells are rendered correctly', () => {
    const data3 = [
      {
        eventType: 'Virus/Malware',
        eventResults: { affectedDevices: 20, detections: 600, unique: 40 },
        selected: true,
      },
    ];

    const columns3 = [
      {
        label: 'Event Type1',
        dataKey: 'eventType',
      },
      {
        label: 'Affected Devices',
        dataKey: 'eventResults.affectedDevices',
      },
      { label: 'Detections', dataKey: 'eventResults.detections' },
      { label: 'Linked field', dataKey: 'eventResults.unique' },
    ];

    const wrapper11 = mount(
      <Table data={data3} columns={columns3} selectable />,
    );
    expect(
      wrapper11
        .find('.rc-checkbox')
        .last()
        .prop('checked'),
    ).toBeTruthy();
  });
});
