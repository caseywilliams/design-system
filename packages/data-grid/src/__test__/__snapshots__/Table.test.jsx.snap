// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test Check component matches previous HTML snapshot 1`] = `
<Table
  className=""
  columns={
    Array [
      Object {
        "columnClassName": "testColumnClassName",
        "dataKey": "eventType",
        "label": "Event Type",
      },
      Object {
        "columnClassName": "testColumnClassName",
        "dataKey": "affectedDevices",
        "label": "Affected Devices",
      },
      Object {
        "columnClassName": "testColumnClassName",
        "dataKey": "detections",
        "label": "Detections",
      },
      Object {
        "columnClassName": "testColumnClassName",
        "dataKey": "Link",
        "label": "Linked field",
      },
    ]
  }
  data={
    Array [
      Object {
        "affectedDevices": 20,
        "detections": 634,
        "eventType": "Virus/Malware",
        "id": 0,
      },
      Object {
        "affectedDevices": 20,
        "detections": 634,
        "eventType": "Spyware/Grayware",
        "id": 1,
      },
      Object {
        "affectedDevices": 15,
        "detections": 598,
        "eventType": "URL Filtering",
        "id": 2,
      },
      Object {
        "affectedDevices": 15,
        "detections": 598,
        "eventType": "Web Reputation",
        "id": 3,
      },
      Object {
        "affectedDevices": 15,
        "detections": 497,
        "eventType": "Network Virus",
        "id": 4,
      },
      Object {
        "affectedDevices": 0,
        "detections": 0,
        "eventType": "Application Control",
        "id": 5,
      },
      Object {
        "affectedDevices": 0,
        "detections": 0,
        "eventType": "Application Control",
        "id": 6,
      },
    ]
  }
  emptyStateHeader="No data available"
  emptyStateMessage="Prompt to action or solution"
  fixed={false}
  fixedColumn={false}
  headerCheckState={false}
  horizontalScroll={false}
  onHeaderChecked={[Function]}
  onRowChecked={[Function]}
  onSort={[Function]}
  rowClassName={[Function]}
  selectable={false}
  sortable={false}
  sortedColumn={
    Object {
      "direction": "",
      "sortDataKey": "",
    }
  }
>
  <div
    className=""
  >
    <table
      className="rc-table"
    >
      <ColumnHeader
        columnHeaderCallBack={[Function]}
        columns={
          Array [
            Object {
              "columnClassName": "testColumnClassName",
              "dataKey": "eventType",
              "label": "Event Type",
            },
            Object {
              "columnClassName": "testColumnClassName",
              "dataKey": "affectedDevices",
              "label": "Affected Devices",
            },
            Object {
              "columnClassName": "testColumnClassName",
              "dataKey": "detections",
              "label": "Detections",
            },
            Object {
              "columnClassName": "testColumnClassName",
              "dataKey": "Link",
              "label": "Linked field",
            },
          ]
        }
        headerCheckState={false}
        onHeaderChecked={[Function]}
        selectable={false}
        sortable={false}
        sortedColumn={
          Object {
            "direction": "",
            "sortDataKey": "",
          }
        }
      >
        <thead>
          <tr
            className="rc-table-header"
          >
            <th
              className="rc-table-header-cell"
              key="eventType"
              onClick={[Function]}
              onKeyPress={[Function]}
              tabIndex={null}
            >
              <span
                as="h6"
                className=""
                color="medium"
              >
                Event Type
              </span>
            </th>
            <th
              className="rc-table-header-cell"
              key="affectedDevices"
              onClick={[Function]}
              onKeyPress={[Function]}
              tabIndex={null}
            >
              <span
                as="h6"
                className=""
                color="medium"
              >
                Affected Devices
              </span>
            </th>
            <th
              className="rc-table-header-cell"
              key="detections"
              onClick={[Function]}
              onKeyPress={[Function]}
              tabIndex={null}
            >
              <span
                as="h6"
                className=""
                color="medium"
              >
                Detections
              </span>
            </th>
            <th
              className="rc-table-header-cell"
              key="Link"
              onClick={[Function]}
              onKeyPress={[Function]}
              tabIndex={null}
            >
              <span
                as="h6"
                className=""
                color="medium"
              >
                Linked field
              </span>
            </th>
          </tr>
        </thead>
      </ColumnHeader>
      <tbody>
        <tr
          className="dg-table-row"
          key="0"
        >
          <td
            className="rc-table-cell testColumnClassName"
            key="eventType"
          >
            Virus/Malware
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="affectedDevices"
          >
            20
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="detections"
          >
            634
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="Link"
          />
        </tr>
        <tr
          className="dg-table-row"
          key="1"
        >
          <td
            className="rc-table-cell testColumnClassName"
            key="eventType"
          >
            Spyware/Grayware
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="affectedDevices"
          >
            20
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="detections"
          >
            634
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="Link"
          />
        </tr>
        <tr
          className="dg-table-row"
          key="2"
        >
          <td
            className="rc-table-cell testColumnClassName"
            key="eventType"
          >
            URL Filtering
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="affectedDevices"
          >
            15
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="detections"
          >
            598
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="Link"
          />
        </tr>
        <tr
          className="dg-table-row"
          key="3"
        >
          <td
            className="rc-table-cell testColumnClassName"
            key="eventType"
          >
            Web Reputation
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="affectedDevices"
          >
            15
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="detections"
          >
            598
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="Link"
          />
        </tr>
        <tr
          className="dg-table-row"
          key="4"
        >
          <td
            className="rc-table-cell testColumnClassName"
            key="eventType"
          >
            Network Virus
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="affectedDevices"
          >
            15
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="detections"
          >
            497
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="Link"
          />
        </tr>
        <tr
          className="dg-table-row"
          key="5"
        >
          <td
            className="rc-table-cell testColumnClassName"
            key="eventType"
          >
            Application Control
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="affectedDevices"
          >
            0
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="detections"
          >
            0
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="Link"
          />
        </tr>
        <tr
          className="dg-table-row"
          key="6"
        >
          <td
            className="rc-table-cell testColumnClassName"
            key="eventType"
          >
            Application Control
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="affectedDevices"
          >
            0
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="detections"
          >
            0
          </td>
          <td
            className="rc-table-cell testColumnClassName"
            key="Link"
          />
        </tr>
      </tbody>
    </table>
  </div>
</Table>
`;

exports[`Snapshot test Check component matches previous HTML snapshot 2`] = `
<Table
  className=""
  columns={
    Array [
      Object {
        "columnClassName": "testColumnClassName",
        "dataKey": "eventType",
        "label": "Event Type",
      },
      Object {
        "columnClassName": "testColumnClassName",
        "dataKey": "affectedDevices",
        "label": "Affected Devices",
      },
      Object {
        "columnClassName": "testColumnClassName",
        "dataKey": "detections",
        "label": "Detections",
      },
      Object {
        "columnClassName": "testColumnClassName",
        "dataKey": "Link",
        "label": "Linked field",
      },
    ]
  }
  data={Array []}
  emptyStateHeader="No data available"
  emptyStateMessage="Prompt to action or solution"
  fixed={false}
  fixedColumn={false}
  headerCheckState={false}
  horizontalScroll={false}
  onHeaderChecked={[Function]}
  onRowChecked={[Function]}
  onSort={[Function]}
  rowClassName={[Function]}
  selectable={false}
  sortable={false}
  sortedColumn={
    Object {
      "direction": "",
      "sortDataKey": "",
    }
  }
>
  <div
    className=""
  >
    <table
      className="rc-table"
    >
      <ColumnHeader
        columnHeaderCallBack={[Function]}
        columns={
          Array [
            Object {
              "columnClassName": "testColumnClassName",
              "dataKey": "eventType",
              "label": "Event Type",
            },
            Object {
              "columnClassName": "testColumnClassName",
              "dataKey": "affectedDevices",
              "label": "Affected Devices",
            },
            Object {
              "columnClassName": "testColumnClassName",
              "dataKey": "detections",
              "label": "Detections",
            },
            Object {
              "columnClassName": "testColumnClassName",
              "dataKey": "Link",
              "label": "Linked field",
            },
          ]
        }
        headerCheckState={false}
        onHeaderChecked={[Function]}
        selectable={false}
        sortable={false}
        sortedColumn={
          Object {
            "direction": "",
            "sortDataKey": "",
          }
        }
      >
        <thead>
          <tr
            className="rc-table-header"
          >
            <th
              className="rc-table-header-cell"
              key="eventType"
              onClick={[Function]}
              onKeyPress={[Function]}
              tabIndex={null}
            >
              <span
                as="h6"
                className=""
                color="medium"
              >
                Event Type
              </span>
            </th>
            <th
              className="rc-table-header-cell"
              key="affectedDevices"
              onClick={[Function]}
              onKeyPress={[Function]}
              tabIndex={null}
            >
              <span
                as="h6"
                className=""
                color="medium"
              >
                Affected Devices
              </span>
            </th>
            <th
              className="rc-table-header-cell"
              key="detections"
              onClick={[Function]}
              onKeyPress={[Function]}
              tabIndex={null}
            >
              <span
                as="h6"
                className=""
                color="medium"
              >
                Detections
              </span>
            </th>
            <th
              className="rc-table-header-cell"
              key="Link"
              onClick={[Function]}
              onKeyPress={[Function]}
              tabIndex={null}
            >
              <span
                as="h6"
                className=""
                color="medium"
              >
                Linked field
              </span>
            </th>
          </tr>
        </thead>
      </ColumnHeader>
      <tbody />
    </table>
    <div
      className="dg-empty-state-container"
    >
      <Heading
        allCaps={false}
        as="h3"
        className="dg-empty-state-header"
        color="medium"
        hero={false}
        label={false}
      >
        <h3
          className="rc-heading rc-heading-h3 rc-heading-medium dg-empty-state-header"
        >
          No data available
        </h3>
      </Heading>
      <Text
        as="div"
        className="dg-empty-state-message"
        color="medium"
        size="medium"
        style={Object {}}
      >
        <div
          className="rc-text rc-text-size-medium rc-text-medium dg-empty-state-message"
          style={Object {}}
        >
          Prompt to action or solution
        </div>
      </Text>
    </div>
  </div>
</Table>
`;
